File: solutions/lua/triangle/triangle.lua
local triangle = {}

function triangle.kind(a, b, c)
    assert(a > 0 and b > 0 and c > 0, "Input Error")
    assert(a + b > c and b + c > a and c + a > b, "Input Error")

    if a == b and b == c then
        return "equilateral"
    elseif a == b or b == c or c == a then
        return "isosceles"
    else
        return "scalene"
    end
end

return triangle

File: solutions/prolog/triangle/triangle.pl
triangle(A, B, C, _) :- 
    (A + B < C; B + C < A; C + A < B), !, fail.

triangle(A, B, C, _) :- 
    A = 0, B = 0, C = 0, !, fail.

triangle(A, A, A, "equilateral") :- !.

triangle(A, B, C, "scalene") :- 
    A \= B, B \= C, C \= A, !.

triangle(A, A, _, "isosceles") :- !.
triangle(A, _, A, "isosceles") :- !.
triangle(_, A, A, "isosceles") :- !.
